### ISSUE TITLE
about gaussian noise and batch size

### ISSUE THREAD

**Issue by gaozhitong** [502463708/Confident_Learning_for_Noisy-labeled_Medical_Image_Segmentation#1]
Hi thank you for your contribution. I have two questions when running the codes.
1. [gaussian noise radius] In noisy dataset generation step, I find the gaussian noise implementation is a little different from the description of edge-distorted in the paper, mainly about the circle radius. 

- In the function: gaussian_noise, the circle's radius is set as a random number from [0, max_radius]. The max_radius, in my understanding, is beta + 3. `radius = np.random.randint(max_radius, size=1)`.  

- In the paper, however, it is said that "edge-distorted was implemented by randomly dilating or eroding the pixels on the boundary with a circle (radius = |β/5 ± 2| pixel)."

  I am quite confused and not sure whether it is a mistake in the code or I set beta wrongly. Following my understanding of the paper's discription, I currently set the label_class_parameter_list as [[5-3, 5+3], [10-3, 10+3], [[15-3, 15+3]] for group A. Pictures after dilating and eroding all look reasonable, but the pictures after edge-distorted are a bit strange. An example is shown below:
![下载 (1)](https://user-images.githubusercontent.com/40078173/100691765-5ce43b80-33c4-11eb-9d8e-2d8680ff132c.png)

2. [batch size] In the config_confident_learning_pixel_level_classification.py, I find the train.batch_size is set as 4. Is it the same setting for you to run the experiments reported in the paper? Or do you use a larger batch_size? 

Thanks very much for your time! Looking forward to your reply.

**Comment by 502463708** [comment]
Thank you so much for your interest in my paper.
For question 1, the edge-distorted label looks correct and the same as I did. I have to admit that there are some issues in the edge-distorted implementation for its failure to resort to the realistic label noise. But it is actually a kind of label noise. You can improve this work by designing the way noise generated close to reality.
 For question 2, the batch size is not that important, you can set it 4, 8 , or 12...

**Comment by 502463708** [comment]
> Hi thank you for your contribution. I have two questions when running the codes.
> 
> 1. [gaussian noise radius] In noisy dataset generation step, I find the gaussian noise implementation is a little different from the description of edge-distorted in the paper, mainly about the circle radius.
> 
> * In the function: gaussian_noise, the circle's radius is set as a random number from [0, max_radius]. The max_radius, in my understanding, is beta + 3. `radius = np.random.randint(max_radius, size=1)`.
> * In the paper, however, it is said that "edge-distorted was implemented by randomly dilating or eroding the pixels on the boundary with a circle (radius = |β/5 ± 2| pixel)."
>   I am quite confused and not sure whether it is a mistake in the code or I set beta wrongly. Following my understanding of the paper's discription, I currently set the label_class_parameter_list as [[5-3, 5+3], [10-3, 10+3], [[15-3, 15+3]] for group A. Pictures after dilating and eroding all look reasonable, but the pictures after edge-distorted are a bit strange. An example is shown below:
>   ![下载 (1)](https://user-images.githubusercontent.com/40078173/100691765-5ce43b80-33c4-11eb-9d8e-2d8680ff132c.png)
> 
> 1. [batch size] In the config_confident_learning_pixel_level_classification.py, I find the train.batch_size is set as 4. Is it the same setting for you to run the experiments reported in the paper? Or do you use a larger batch_size?
> 
> Thanks very much for your time! Looking forward to your reply.

**Comment by gaozhitong** [comment]
Thank you so much for the speedy reply. I am now in the confident map generation step, and I have some understanding of your pipline but not very sure, which I want to get your confirmation.

1. It seems that you divide the whole dataset into two parts: sub1 and sub2, and train two networks (net1 and net2) individually on these two subsets. Then in the confident map generation step, you correct sub1 using net2 and correct sub2 using net1. Am I understanding right? 
I think this is a good idea but I do not see relevant design described in the paper. So, is this an add-on after you finish the paper or the experiments on the original paper actually does so?

2. In this confident map generation step, should I use the best ckpt or the ckpt at 1000 epoch?

Looking forward to your reply!

**Comment by 502463708** [comment]
Since it is my first time to make my code open-sourced, I have already imagined someone could get confused about my code, LOL. 
For Q1:
But you exhibit a good ability to read code and have a correct understanding of it. Yes, because the CNN could overfit the noisy label, I divided the training set into sub1 and sub2, and characterize the label noises with net2 and net1, inversely. And then the confident map of the whole training set can be generated.
This is a trick to avoid overfitting the noisy labels, and I did not describe it in my paper, because I think it is not the key insight of my paper and I am afraid that my paper would get more complex. 
For Q2:
You can use the best ckpt.

**Comment by gaozhitong** [comment]
Hi,  I have finished running the pipline😜. Thanks again, I couldn’t have pulled this off without you.

The last thing I want to confirm is whether to use the best ckpt or the 1000 epoch ckpt for student net to compare with the results of "Baseline + CL+SLSR" shown in the paper.

Looking forward to your reply!

**Comment by 502463708** [comment]
Yes, you can probably reproduce a comparable result.
But the figures you reproduce might not be the exact same as those in the paper due to the random training process.

**Comment by gaozhitong** [comment]
Yes, it is a little hard considering random training process and random noisy generation.
My situation now is the best ckpt 's score is larger than the figures reported in the paper, and the 1000 epoch ckpt's score is smaller than the figures reported in the paper. I want to know what you use in the paper, best ckpt 's score or 1000 epoch ckpt's score.
Thanks!

**Comment by 502463708** [comment]
I reported the figures evaluated by the best ckpt model.
Thanks for your interest, and could you please share something about yourself with me if you do not mind?
Something like where are you originally from? Which institute do you work in? Whether you are preparing for a MICCAI 2021 paper? My email address is minqing_zhang@163.com.
